rfc/router-storage.md

```diff
@@ -0,0 +1,767 @@
+# Router.Storage
+
+## Authors
+
+- [hannah]
+
+## Abstract
+
+Proposes extensions to the [W3 Blob Protocol](https://github.com/web3-storage/specs/blob/main/w3-blob.md) to enable users to expess preferences about how their blobs are stored, and a suggested implementation for a W3 blob "router", a capability provider for `/space/content/add/blob` that would invoke `/service/blob/allocate` and `service/blob/accept` on different storage nodes, based on the users preferences
+
+## W3 Blob Protocol
+
+To enable users to express intent for how they would like data stored, an optional `storagePreferences` argument is added to the `space/content/add/blob` invocation:
+
+```ts
+type AddBlob = {
+  cmd: "/space/content/add/blob"
+  sub: SpaceDID
+  args: {
+    blob: Blob
+    storagePreferances?: StoragePreferances
+  }
+}
+
+// intentionally left blank for now to expand over time
+type StoragePreferences = {
+}
+
+type Blob = {
+  digest:   Multihash
+  size:     int
+}
+
+type Multihash = bytes
+type SpaceDID = string
+```
+
+Expecting that well supported preferences will evolve over time, for now we are not attempting to define specific preference properties.
+
+## Routed implementation of `space/content/add/blob`
+
+A blob router provides the `space/content/add/blob` capability by routing subsequent effects to seperate storage nodes providing `service/blob/allocate` and `service/blob/accept` capabilities.
+
+The blob router contains a state of storage nodes available.
+
+```ts
+type StorageNode = {
+  did: ProviderDID
+  w3upEndpoint: URL // a UCAN invocation endpoint providing at minimum `service/blob/allocate` and `service/blob/accept`
+  storageProperties?: StorageProperties
+}
+
+type StorageNodes = {
+  [ProviderDID]: StorageNode
+}
+
+// intentionally left blank for now to expand over time
+type StorageProperties = {
+}
```

Comment:
So what we have coming in UCAN 1.0 and something I already have implemented (but not shipped) is UCAN policy system which is roughly equivalent of WHERE clause in SQL, so when I say `StorageCapibaliites` I imply something like

```ts
{ cmd: '/memory/allocate',
  with: did:...,
  where: [
    ["<=", ".blob.size", 1024]], // size of the blob should be < 1024
    ["some", ["asia", "europe"], [= ".region" "."]] // reigion should be aisa or europe
  ]
}
```


As mentioned benefit of using UCAN policy system would be that it would be well defined and will not require us to roll out custom code for figuring out which node providers satisfy desired criteria as we would be able to reuse most logic. It will also give us framework for expressing properties, which will be limiting but in my experience good guardrails to have.